### DISCUSSION TITLE
Allow "Terms and Conditions" after every login.

### DISCUSSION THREAD

**Discussion by mddamato** [keycloak/keycloak#9149] [Ideas]
### Description

The ability to set "terms and conditions" to be accepted after every login, not just a one time requiredAction.

### Motivation

Many organizations including government organizations require a notice to be accepted by users every time they login. The terms and conditions feature of Keycloak is a good solution to this because it allows administrators to offset this requirement from the application UI to Keycloak and can therefore be reused across many applications within the organization. This also provides better auditing and traceability for legal reasons.

**Comment by SergiiKram** [comment]
I also see it as a requirement more and more often.

**Comment by mddamato** [comment]
example requirement definition: https://www.stigviewer.com/stig/application_server_security_requirements_guide/2019-09-30/finding/V-35096

This would still be nice to have 2 years later.

**Comment by ianwallen** [comment]
We got it working by creating a custom Authenticator.

It will always add the terms and conditions as a required action right after the user login. This will ensure that users are always prompted to accept the terms and condition after they login.

Here is the code for the new Authenticator. You will need to look into keycloak development on how to create a custom SPI.

```
import org.keycloak.authentication.AuthenticationFlowContext;
import org.keycloak.authentication.authenticators.resetcred.AbstractSetRequiredActionAuthenticator;
import org.keycloak.models.UserModel;

public class TermAndConditionsAuthenticator  extends AbstractSetRequiredActionAuthenticator {
    public static final String PROVIDER_ID = "terms-and-conditions";

    @Override
    public void authenticate(AuthenticationFlowContext context) {
        if (context.getExecution().isRequired()) {
            context.getAuthenticationSession().addRequiredAction(UserModel.RequiredAction.TERMS_AND_CONDITIONS);
        }
        context.success();
    }

    @Override
    public String getDisplayType() {
        return "Sets the Terms and Conditions required action";
    }

    @Override
    public String getHelpText() {
        return "Sets the Terms and Conditions required action if execution is REQUIRED.";
    }

    @Override
    public String getId() {
        return PROVIDER_ID;
    }
}
```
Ensure that it is deployed in provider info

![image](https://github.com/keycloak/keycloak/assets/1868233/7c4159f7-7cc8-4c65-82d6-ceab54acda67)

![image](https://github.com/keycloak/keycloak/assets/1868233/175010fc-edee-4408-847e-52bdbfe007c0)

Then make a copy of the browser flow from the authentication section. And add the new step in the browser form and select the new terms and conditions option.

![image](https://github.com/keycloak/keycloak/assets/1868233/bfc5e4d9-823c-44b8-ba46-5a9a1f6f7909)

Once complete you should see something like the following.

![image](https://github.com/keycloak/keycloak/assets/1868233/d5a39694-eeaa-4a50-84b1-ac2af4e5f74f)

**Reply by mddamato** [reply]
Thank you for sending this @ianwallen , I will give this a shot next week!

**Reply by vigkastHV** [reply]
@ianwallen , Can you please share the full code repository for this custom terms and conditions authenticator? I have similar requirement like this. It would be great if you help me with this.

**Reply by ianwallen** [reply]
@vigkastHV 
I is currently in a private repository so I cannot share.

This should be similar to implementing any SPI which keycloak has documented here
https://www.keycloak.org/docs/latest/server_development/#_auth_spi
